### PR TITLE
SMS user registration alerts

### PR DESCRIPTION
**Pull request by proteusvacuum**
## Product Description

This allows admins to add a list of emails that will be alerted whenever a new mobile worker is registered via SMS. 

Screenshot from 2021-11-19 12-18-21

https://docs.google.com/document/d/1qgKxA-xCvhvSnbrZy4scSbmjOsZA9KYgpbqBAce01sI/edit#heading=h.ry3kp26qtvh9
https://dimagi-dev.atlassian.net/browse/SC-1827

## Technical Summary

This adds a new domain property which stores a list of emails. If that list is present, an email alert is sent whenever a new user is registered via SMS. 

### Safety story

This area seems generally well tested.

### Automated test coverage
Modified and added some tests. 

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by mkangia** [comment]
> If a user changes their SIM card, a new user needs to be created 

@proteusvacuum can you share why this is needed? 
I understand that the user needs to be assigned to a location, so we notify admins for a new worker, so that they can do so.
In case of number change, They can have a process to request the admin to update/add their phone number. Why do they need to register again from the new number?

File: corehq/apps/sms/api.py [hunk 1]
```diff
@@ -458,31 +492,12 @@ def process_sms_registration(msg):
...
-                        entry.save()
-                        registration_processed = True
-
-                        if domain_obj.enable_registration_welcome_sms_for_mobile_worker:
-                            send_sms(domain_obj.name, None, cleaned_phone_number,
-                                     get_message(MSG_REGISTRATION_WELCOME_MOBILE_WORKER, domain=domain_obj.name))
-                    except ValidationError as e:
-                        send_sms(domain_obj.name, None, cleaned_phone_number, e.messages[0])
+                    registration_processed = register_sms_user(
+                        username=cleaned_phone_number if keyword4 == '' else keyword4,
```
Comment (mkangia) [inline corehq/apps/sms/api.py]: nit: I see we have flipped the logic here which should be safe but just scares me a bit with comparisons like None with an empty string. 
`username=keyword4 if keyword4 != '' else cleaned_phone_number` would be the exact same.

File: corehq/apps/sms/tests/test_registration.py [hunk 2]
```diff
@@ -52,6 +51,7 @@ def tearDown(self):
         delete_domain_phone_numbers(self.domain)
         SQLMobileBackendMapping.unset_default_domain_backend(self.domain)
         self.backend.delete()
+        delete_all_users()
```
Comment (mkangia) [inline corehq/apps/sms/tests/test_registration.py:55]: :+1:

File: corehq/apps/sms/tests/test_registration.py [hunk 3]
```diff
@@ -27,8 +25,9 @@ class RegistrationTestCase(BaseSMSTest):
 
     def setUp(self):
         super(RegistrationTestCase, self).setUp()
+        delete_all_users()
 
-        self.domain = 'sms-reg-test-domain'
+        self.domain = 'sms-reg-test-domain-1'
```
Comment (mkangia) [inline corehq/apps/sms/tests/test_registration.py:31]: why was this needed? and also the "delete_all_users" in beginning. Are these left overs while writing tests?

File: corehq/apps/sms/forms.py [hunk 4]
```diff
@@ -616,6 +634,17 @@ def __init__(self, data=None, domain=None, is_previewer=False, *args, **kwargs):
             "add_window_method": "addSMSConversationTime",
         }
 
+    def set_admin_email_choices(self, data, kwargs):
+        email_choices = []
+        if 'initial' in kwargs and 'sms_worker_registration_alert_emails' in kwargs['initial']:
+            email_choices = kwargs['initial']['sms_worker_registration_alert_emails']
+        elif data:
+            email_choices = data.getlist('sms_worker_registration_alert_emails')
```
Comment (mkangia) [inline corehq/apps/sms/forms.py:642]: is this if/else for GET and POST?

**Review by mkangia — COMMENTED** [review]
Looks good code wise. Just couple of queries.

Quick question, why are we showing the message content in email? Isn't that the standard "Join XYZ" message.

Also, Who has access to set these notifications? Just thinking of the security aspect of this.
We are sharing a user's phone number to another user over email. So if this notification is going to a user who already has access to the new user's edit page, then it sounds okay but not otherwise.

File: corehq/apps/sms/forms.py (hunk 4 above)
Comment (proteusvacuum) [reply corehq/apps/sms/forms.py:642]: Yes. Is there a better way of doing this in a form?

File: corehq/apps/sms/tests/test_registration.py (hunk 3 above)
Comment (proteusvacuum) [reply corehq/apps/sms/tests/test_registration.py:31]: Whoops, yes I left this in by accident.

**Comment by proteusvacuum** [comment]
> Quick question, why are we showing the message content in email? Isn't that the standard "Join XYZ" message.

You can add other keywords in the message. It is possible that a user could add a message for an admin, for example. This is so projects can add different requirements for sign up.

> Also, Who has access to set these notifications? Just thinking of the security aspect of this.

This requires the "edit messaging" permission. 

> We are sharing a user's phone number to another user over email. So if this notification is going to a user who already has access to the new user's edit page, then it sounds okay but not otherwise.

Hmm, are you saying this should be limited to only web users who have admin (or user edit) privileges on HQ? And if that user or permission is revoked, we should remove that email from this list? I do kind of agree and hadn't though of that, although it adds quite a bit of scope to this feature. 
What if I just changed the message so that it doesn't show the mobile number, username or message, and just presents a link to the user edit page?

File: corehq/apps/sms/forms.py (hunk 4 above)
Comment (mkangia) [reply corehq/apps/sms/forms.py:642]: Was just trying to understand why there is a branching here.

Checking that if there are initial values or not, does not reveal the actual need for the check for request.
I was thinking of
```
if request.method == 'GET':
    email_choices = kwargs['initial']['sms_worker_registration_alert_emails']
elif request.method == 'POST' and data:
    email_choices = data.getlist('sms_worker_registration_alert_emails')
```
unless that breaks the forms.
Though, I am fine with the current implementation as well.

**Comment by mkangia** [comment]
> You can add other keywords in the message. It is possible that a user could add a message for an admin, for example

Communicating with admin/email notifiers through registration SMS sounds sneaky.
Cool. I was thinking if someone needs to see the SMS content, they can just visit HQ. Can we just share a link for the same instead? Having the content in the email itself would be handy. And if it's useful here that is okay then. I guess sharing sms content that user sent to register themselves doesn't come as close as phone number from security perspective.

**Comment by proteusvacuum** [comment]
> I was thinking if someone needs to see the SMS content, they can just visit HQ

Can you elaborate? I don't think we currently have a way of viewing a single inbound SMS message, right? So what do you imagine this workflow would be - a link to the messaging history report, and the user would then scroll through to find the message they want?

**Comment by mkangia** [comment]
>  I don't think we currently have a way of viewing a single inbound SMS message, right?

Been a while I interacted with SMS app. But there should be a view of the SMS or just the logged event that we can link to? Do you have a domain on staging where we can check incoming messages in report?

**Comment by proteusvacuum** [comment]
Hey @mkangia from my understanding, there is no basic view of an SMS. AFAICT `MessagingEvent`s don't get created for registration SMS messages. 

Can I just understand what you think the threat model of this is?
- Only people who can see SMS messages can add email addresses here. In theory they will add trusted entities. If they don't they had access to SMS messages anyway, so there is no new threat here.
- SMS messages are sent unencrypted (since SMS is an unencrypted medium). Forwarding this text via email to a trusted party also doesn't change anything here.
- I've removed the mobile number from the email, although given the two points above, I can't figure out why having that in an email would be a problem.

File: corehq/apps/sms/forms.py (hunk 4 above)
Comment (Charl1996) [reply corehq/apps/sms/forms.py:642]: @proteusvacuum I must agree with @mkangia ; it's not clear why the branching exists. I'd even be satisfied with a comment if the `request.method == 'GET'` is too hard to get in.

File: corehq/apps/sms/forms.py [hunk 5]
```diff
@@ -381,6 +390,14 @@ def section_registration(self):
                     "[project] worker [username]' (where [project] is your "
                     " project space and [username] is an optional username)"
                     ", and the system will add them as a mobile worker."),
+                data_bind="value: sms_mobile_worker_registration_enabled",
+            ),
+            crispy.Div(
+                hqcrispy.FieldWithHelpBubble(
+                    "sms_worker_registration_alert_emails",
+                    help_bubble_text=_("Email these people when new users sign register through SMS"),
```
Comment (kaapstorm) [inline corehq/apps/sms/forms.py:398]: Are you using "register" as a noun, like "sign the register", or did you mean to change "sign up" to "register"?

```suggestion
                    help_bubble_text=_("Email these people when new users register through SMS"),
```

File: corehq/apps/sms/forms.py (hunk 4 above)
Comment (kaapstorm) [reply corehq/apps/sms/forms.py:642]: It appears from [Django docs](https://docs.djangoproject.com/en/2.2/ref/forms/api/#django.forms.Form.has_changed) that a form can be instantiated with both `data` and `initial`. Maybe that `elif data:` should be an `if data:`?

**Review by kaapstorm — APPROVED** [review]

**Comment by Charl1996** [comment]
@proteusvacuum Are we handling permission revokes and user deletions in a subsequent ticket?

@mkangia I must agree with Farid; I see no risk in showing the SMS message text in the email?

> AFAICT MessagingEvents don't get created for registration SMS messages.

This sounds like something that should be happening IMHO (but in another ticket).

**Review by Charl1996 — APPROVED** [review]

**Comment by kaapstorm** [comment]
> Why do they need to register again from the new number?

My (very limited) understanding is that these mobile workers are offline 100% of the time. Registering via SMS is how they notify admins (whose contact details they probably don't have) that their number has changed.

Regarding notifications being sent to **ex**-admins, I agree that's something we'd want to avoid, and feels sloppy. Linking to the user edit page seems like a reasonable compromise to me.

File: corehq/apps/sms/tests/test_registration.py [hunk 6]
```diff
@@ -80,18 +80,26 @@ def assertLastOutgoingSMS(self, phone_number, expected_texts):
...
 
         # Test with mobile worker registration enabled
         self.domain_obj.sms_mobile_worker_registration_enabled = True
         self.domain_obj.save()
 
-        incoming('+9991234567', 'JOIN {} WORKER tester'.format(self.domain), self.backend.hq_api_id)
+        incoming(phone_number, 'JOIN {} WORKER tester'.format(self.domain), self.backend.hq_api_id)
         self.assertIsNotNone(CommCareUser.get_by_username(formatted_username))
 
+        # Test with no username
```
[the remark this answers is not in the document]
Comment (proteusvacuum) [reply corehq/apps/sms/tests/test_registration.py:106]: I agree, updated.

File: corehq/apps/sms/forms.py (hunk 5 above)
Comment (proteusvacuum) [reply corehq/apps/sms/forms.py]: 🦅 👁️

**Comment by mkangia** [comment]
Thanks for your input @Charl1996 @kaapstorm 

@proteusvacuum 
So looks like, Originally, the SMS is only available to the user sending it and HQ (where it is not saved), so the content gets lost.
We are relying on the project space to ensure the notification is sent to legit users, who they trust, which is my concern here. 
If there is custom content in the registration SMS sent, there is no guarantee what it will be.
I might be overthinking this, scenario I can think of is someone leaves the company (the domain belongs to), there email is left in the list and they keep getting notified about the state of the project and the project comes to us asking why that happened even when they revoked their access to project. It's not just the content of the SMS itself but just general project information/progress that is going out.
Can we?
1) check that these users are web users of the domain when we notify each time
2) when the email is added we can ensure they are entering the correct email of an active web user or go fancy and let them search active web users when adding these
I feel those two points will rest my concerns.

Feel free to do a quick check with Clayton if this seems like an overkill. Looks like we are in a split here.

File: corehq/apps/sms/templates/sms/email/new_sms_user.html [hunk 7]
```diff
@@ -4,10 +4,6 @@
     A new user {{ username }} registered for project {{ domain }} through SMS.
   </p>
 
-  <p>
-    Here is the full content of their registration message: {{ message_text }}
-  </p>
-
```
Comment (mkangia) [inline corehq/apps/sms/templates/sms/email/new_sms_user.html]: :+1:

Comment (proteusvacuum) [reply corehq/apps/sms/templates/sms/email/new_sms_user.html]: @mkangia do you think this is a sufficient change? I can add an "SMS view" later if it is needed.

Comment (mkangia) [reply corehq/apps/sms/templates/sms/email/new_sms_user.html]: Yup. I feel good to wait on adding the SMS view, until its requested

Comment (proteusvacuum) [reply corehq/apps/sms/templates/sms/email/new_sms_user.html]: Great! If so, would you be OK approving this PR then?

**Comment by proteusvacuum** [comment]
Given Manish's approval of the last commit, and the approvals from others, I'm going to merge this.

File: corehq/apps/sms/templates/sms/email/new_sms_user.html (hunk 7 above)
Comment (mkangia) [reply corehq/apps/sms/templates/sms/email/new_sms_user.html]: oh sorry, I was waiting for your response on other comments though, which were non-blocking.
